common/src/java/org/apache/hive/common/util/ZooKeeperHiveHelper.java

```diff
@@ -0,0 +1,237 @@
+// including an instance of first class in the second and the third, but there's isn't much stuff
+// in the first and the third.. Also note that the third functionality overlaps with
+// ZooKeeperHiveClientHelper class, but that overlap is very small. So for now all the three
+// functionality are bundled in a single class.
+public class ZooKeeperHiveHelper {
+  public static final Logger LOG = LoggerFactory.getLogger(ZooKeeperHiveHelper.class.getName());
+  public static final String ZOOKEEPER_PATH_SEPARATOR = "/";
+
+  private String quorum = null;
+  private String clientPort = null;
+  private String rootNamespace = null;
+  private boolean deregisteredWithZooKeeper = false; // Set to true only when deregistration happens
+  private int sessionTimeout;
+  private int baseSleepTime;
+  private int maxRetries;
+
+  private CuratorFramework zooKeeperClient;
+  private PersistentEphemeralNode znode;
+
+  public ZooKeeperHiveHelper(String quorum, String clientPort, String rootNamespace,
+                             int sessionTimeout, int baseSleepTime, int maxRetries) {
+    // Get the ensemble server addresses in the format host1:port1, host2:port2, ... . Append
+    // the configured port to hostname if the hostname doesn't contain a port.
+    String[] hosts = quorum.split(",");
+    StringBuilder quorumServers = new StringBuilder();
+    for (int i = 0; i < hosts.length; i++) {
+      quorumServers.append(hosts[i].trim());
+      if (!hosts[i].contains(":")) {
+        quorumServers.append(":");
+        quorumServers.append(clientPort);
+      }
+
+      if (i != hosts.length - 1)
+        quorumServers.append(",");
+    }
+
+    this.quorum = quorumServers.toString();
+    this.clientPort = clientPort;
+    this.rootNamespace = rootNamespace;
+    this.sessionTimeout = sessionTimeout;
+    this.baseSleepTime = baseSleepTime;
+    this.maxRetries = maxRetries;
+  }
+
+  /**
+   * Get the ensemble server addresses. The format is: host1:port, host2:port..
+   **/
+  public String getQuorumServers() {return quorum;}
+
+  /**
+   * Adds a server instance to ZooKeeper as a znode.
+   *
+   * @throws Exception
+   */
+  public void addServerInstanceToZooKeeper(String znodePathPrefix, String znodeData,
+                                           ACLProvider zooKeeperAclProvider,
+                                           ZKDeRegisterWatcher watcher) throws Exception {
+    // This might be the first server getting added to the ZooKeeper, so the parent node may need
+    // to be created.
+    zooKeeperClient = startZookeeperClient(zooKeeperAclProvider, true);
+
+    // Create a znode under the rootNamespace parent for the given path prefix for a server. Also
+    // add a watcher to watch the znode.
+    try {
+      String pathPrefix = ZOOKEEPER_PATH_SEPARATOR + rootNamespace
+                      + ZOOKEEPER_PATH_SEPARATOR + znodePathPrefix;
+      byte[] znodeDataUTF8 = znodeData.getBytes(Charset.forName("UTF-8"));
```

Comment:
use can use StandardCharsets.UTF_8 directly too.

Reply:
This is original code which I used as is during refactoring. But I with your suggestion, we save a few cycles. So, done.